lib/Sema/CSApply.cpp

```diff
@@ -413,43 +414,67 @@ namespace {
             if (conformance.isConcrete()) {
               if (auto witness =
                       conformance.getConcrete()->getWitnessDecl(decl)) {
-                // Hack up an AST that we can type-check (independently) to get
-                // it into the right form.
-                // FIXME: the hop through 'getDecl()' is because
-                // SpecializedProtocolConformance doesn't substitute into
-                // witnesses' ConcreteDeclRefs.
                 Type expectedFnType = simplifyType(overload.openedType);
                 assert(expectedFnType->isEqual(
                            fullType->castTo<AnyFunctionType>()->getResult()) &&
                        "Cannot handle adjustments made to the opened type");
+
+                // Build the AST for the call to the witness.
+                //
+                // We have to recover substitutions in this hacky way because
+                // the AST does not retain enough information to devirtualize
+                // calls like this.
+                auto witnessType = witness->getInterfaceType();
+
+                // Compute the substitutions.
+                SubstitutionMap subMap;
+                if (auto *gft = witnessType->getAs<GenericFunctionType>()) {
+                  auto sig = gft->getGenericSignature();
+                  auto *env = sig->getGenericEnvironment();
+
+                  witnessType = FunctionType::get(gft->getParams(),
+                                                  gft->getResult(),
+                                                  gft->getExtInfo());
+                  witnessType = env->mapTypeIntoContext(witnessType);
+
+                  TypeSubstitutionMap subs;
+                  witnessType->substituteBindingsTo(
+                    fullType,
```

Comment:
What if the witness is a top-level function? I think we may have to strip the curried self params for them to line up correctly.

Reply:
That explains why the stdlib failed to build. Thanks!